vNext/UnityGLTFv2.cs

```diff
@@ -0,0 +1,181 @@
+    ColliderType Collider { get; set; }
+}
+
+public partial class UnityGLTFExporter
+{
+    /// <param name="dataWriter">Interface for handling the streams of data to write out</param>
+    /// <param name="exportConfig">Configuration of extension settings and export otpions</param>
+    public UnityGLTFImporter(
+        IDataWriter dataWriter,
+        UnityGLTFExporterConfig exportConfig = new UnityGLTFExporterConfig()
+        );
+
+    /// <summary>
+    /// Exports a Unity object to a glTF file
+    /// </summary>
+    /// <param name="unityObject">The object to export</param>
+    /// <returns></returns>
+    public Task<GLTFFObject> Export(
+        GameObject unityObject
+        );
+}
+
+/// <summary>
+/// Writes data for an export operation
+/// </summary>
+public class IDataWriter
+{
+    Task<bool> WriteStream(string uri, Stream stream);
+}
+
+public class UnityGLTFExporterConfig
+{
+    public UnityGLTFExporterConfig();
+    public UnityGLTFExporterConfig(GLTFExtensionRegistry registry, GLTFExportOptions exportOptions);
+}
+
+public class GLTFExportOptions
+{
+    /// <summary>Whether to write the object out as a GLB</summary>
+    bool ShouldWriteGLB { get; set ; }
+}
+
+/// <summary>
+/// Unity wrapper for glTF object schema class from GLTFSerialization
+/// Properly cleans up data
+/// </summary>
+public class UnityGLTFObject : IDisposable
+{
+    /// <summary>
+    /// Constructor for already parsed glTF or GLB
+    /// </summary>
+    /// <param name="gltfObject">Already parsed glTF or GLB</param>
+    public UnityGLTFObject(IGLTFObject gltfObject);
+
+    /// <summary>
+    /// Constructor for not yet parsed glTF or GLB
+    /// The IDataReader will handle loading the data to load the file
+    /// </summary>
+    /// <param name="fileName">Name of file to load</param>
+    public UnityGLTFObject(string fileName);
+
+    internal AssetCache AssetCache { get; }
+}
+
+/// <summary>
+/// Unity glTF extension wrapper
+/// </summary>
+public interface IUnityGLTFExtension
+{
+    IGLTFExtension GLTFExtension { get; };
+
+    Task<ExtensionReturnObject<GameObject>> CreateSceneAsync(GLTF.Schema.GLTFScene gltfScene, int indexToLoad);
+    Task<ExtensionReturnObject<GameObject>> CreateNodeAsync(Node gltfNode, int indexToLoad);
+    Task<ExtensionReturnObject<MeshPrimitive>> CreateMeshPrimitiveAsync(GLTFMesh mesh, ILoaderContext loaderContext, int indexToLoad);
+    /// etc. 
+}
+
+public struct ExtensionReturnObject<T>
+{
+    ExtensionContinuationBehavior ContinuationBehavior;
+    T ReturnObject;
+}
+
+public enum ExtensionContinuationBehavior
+{
+    NotHandled,
+    ContinueDefaultExecution,
+    Handled
+}
+
+// Example calling pattern:
+public async void LoadGLBs()
+{
+    UnityGLTFObject sampleObject = new UnityGLTFObject("http://samplemodels/samplemodel.glb");
+    UnityGLTFObject boxObject = new UnityGLTFObject("http://samplemodels/box.glb");
+    IDataLoader dataLoader = new WebRequestLoader();

```

Comment:
Couldn't we just read the binary data asynchronously from disk?

Reply:
Yes you can specify a FileLoader if desired

Reply:
Let's not assume disk access. We're pulling it from the network.